src/backend/distributed/executor/multi_utility.c

```diff
@@ -2921,6 +2930,142 @@ ErrorIfUnsupportedTruncateStmt(TruncateStmt *truncateStatement)
 }
 
 
+/*
+ * ProcessTruncateStatement handles distributed locking
+ * of truncated tables before standard utility takes over.
+ *
+ * Actual distributed truncation occurs inside truncate trigger.
+ *
+ * This is only for distributed serialization of truncate commands.
+ * The function assumes that there is no foreign key relation between
+ * non-distributed and distributed relations.
+ */
+static void
+ProcessTruncateStatement(TruncateStmt *truncateStatement)
+{
+	LockTruncatedRelationMetadataInWorkers(truncateStatement);
```

Comment:
I think one side effect of acquiring locks before process utility is that even if there'll be some errors, we'll send the commands. However, that seems OK, right?
An example:
```SQL
postgres=# TRUNCATE users_table;
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 351, '2018-09-21 11:05:28.847309+03');
DETAIL:  on server localhost:9700
LOG:  issuing SELECT lock_relation_if_exists('public.users_table', 'ACCESS EXCLUSIVE');
DETAIL:  on server localhost:9700
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 351, '2018-09-21 11:05:28.847309+03');
DETAIL:  on server localhost:9701
LOG:  issuing SELECT lock_relation_if_exists('public.users_table', 'ACCESS EXCLUSIVE');
DETAIL:  on server localhost:9701
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 351, '2018-09-21 11:05:28.847309+03');
DETAIL:  on server localhost:9702
LOG:  issuing SELECT lock_relation_if_exists('public.users_table', 'ACCESS EXCLUSIVE');
DETAIL:  on server localhost:9702
LOG:  issuing SELECT lock_relation_if_exists('public.events_table', 'ACCESS EXCLUSIVE');
DETAIL:  on server localhost:9700
LOG:  issuing SELECT lock_relation_if_exists('public.events_table', 'ACCESS EXCLUSIVE');
DETAIL:  on server localhost:9701
LOG:  issuing SELECT lock_relation_if_exists('public.events_table', 'ACCESS EXCLUSIVE');
DETAIL:  on server localhost:9702
LOG:  issuing ROLLBACK
DETAIL:  on server localhost:9700
LOG:  issuing ROLLBACK
DETAIL:  on server localhost:9701
LOG:  issuing ROLLBACK
DETAIL:  on server localhost:9702
ERROR:  cannot truncate a table referenced in a foreign key constraint
DETAIL:  Table "events_table" references "users_table".
HINT:  Truncate table "events_table" at the same time, or use TRUNCATE ... CASCADE.
```

Reply:
once you hit `ERROR` transaction is rolled back and locks are released.